hivemind/utils/mpfuture.py

```diff
@@ -2,171 +2,260 @@
+    def _state(self) -> State:
+        shared_state = ALL_STATES[self._shared_state_code.item()]
+        return self._state_cache.get(shared_state, shared_state)
+
+    @_state.setter
+    def _state(self, new_state: State):
+        self._shared_state_code[...] = ALL_STATES.index(new_state)
+        if self._state in TERMINAL_STATES and self._loop is not None and not self._aio_event.is_set():
+            self._set_event_threadsafe()
+
+    def _set_event_threadsafe(self):
         try:
-            self._recv_updates(timeout=0)
-        except TimeoutError:
-            pass
+            loop = asyncio.get_running_loop()
+        except RuntimeError:
+            loop = None
+
+        async def _event_setter():
+            self._aio_event.set()
+
+        if loop == self.get_loop():
+            asyncio.create_task(_event_setter())
+        else:
+            asyncio.run_coroutine_threadsafe(_event_setter(), self._loop)
+
+    @property
+    def _sender_pipe(self) -> mp.connection.Connection:
+        """ a pipe that can be used to send updates to the MPFuture creator """
+        return self.global_mpfuture_senders[self._origin_pid]
+
+    @classmethod
+    def _initialize_mpfuture_backend(cls):
+        pid = os.getpid()
+        logger.debug(f"Initializing MPFuture backend for pid {pid}")
+        assert pid != cls.active_pid and pid not in cls.global_mpfuture_senders, "already initialized"
+
+        recv_pipe, send_pipe = mp.Pipe(duplex=False)
+        cls.active_pid, cls.active_futures, cls.global_mpfuture_senders[pid] = pid, {}, send_pipe
+        cls.pipe_waiter_thread = threading.Thread(target=cls._process_updates_in_background, args=[recv_pipe],
+                                                  name=f'{__name__}.BACKEND', daemon=True)
+        cls.pipe_waiter_thread.start()
+
+    @classmethod
+    def _process_updates_in_background(cls, receiver_pipe: mp.connection.Connection):
+        pid = os.getpid()
+        while True:
+            try:
+                uid, update_type, payload = receiver_pipe.recv()
+                if uid not in cls.active_futures:
+                    logger.debug(f"Ignoring update to future with uid={uid}: the future is no longer active.")
+                elif update_type == UpdateType.RESULT:
+                    cls.active_futures.pop(uid).set_result(payload)
+                elif update_type == UpdateType.EXCEPTION:
+                    cls.active_futures.pop(uid).set_exception(payload)
+                elif update_type == UpdateType.CANCEL:
+                    cls.active_futures.pop(uid).cancel()
+                else:
+                    raise RuntimeError(f"MPFuture received unexpected update type {update_type}")
+            except (BrokenPipeError, EOFError):
+                logger.debug(f"MPFuture backend was shut down (pid={pid}).")
+            except Exception as e:
+                logger.warning(f"Internal error (type={e}, pid={pid}): could not retrieve update for MPFuture.")
```

Comment:
```suggestion
                logger.error(f"Internal error (type={e}, pid={pid}): could not retrieve update for MPFuture.")
```

nit: Changing this to match the log level of `logger.exception(e)` below.